agent/internal/agent.go

```diff
@@ -181,12 +184,79 @@ func (a *agent) handleAPIRequest(ctx *actor.Context, apiCtx echo.Context) {
 	}
 }
 
+// readCertPool constructs a pool of certificates from the given file (which should contain only a
+// sequence of PEM-encoded certificates). It overlaps with CertPool.AppendCertsFromPEM in
+// functionality, but that function is excessively flexible in what it accepts, so we choose not to
+// use it.
+func readCertPool(file string) (*x509.CertPool, error) {
+	data, err := ioutil.ReadFile(file) //nolint:gosec
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to read certificate file")
+	}
+	data = bytes.TrimSpace(data)
+	if len(data) == 0 {
+		return nil, errors.New("certificate file contains no data")
+	}
+
+	pool := x509.NewCertPool()
+	for len(data) > 0 {
+		var block *pem.Block
+		block, data = pem.Decode(data)
+
+		if block == nil {
+			if len(data) > 0 {
+				return nil, errors.New("extra data in certificate file")
```

Comment:
question: is a `pem` file strictly invalid if it contains any extra data? I'm not an expert, just the API for `pem.Decode` (scan for the next PEM formatted block and skip everything else) makes me think the answer may be no.

I see `CertPool.AppendCertsFromPEM` does not have this check and that it will ignore PEM formatted blocks that are not certificates. If `pem` files with those qualities aren't strictly invalid, can we just use it and accept them?

Reply:
Okay, turns out everything (well, at least the things I tried, `curl` and `openssl`) does just accept arbitrary data around the PEM blocks in a file. That being the normal behavior sounded so strange to me that I didn't even consider the possibility. `AppendCertsFromPEM` is more lenient still and will power through files that the other two reject (with, e.g. mismatched BEGIN/END lines), but at that point I guess it's close enough to just run with it.